scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/EvaluableLazyParameter.scala

```diff
@@ -14,19 +20,88 @@ class EvaluableLazyParameter[T <: AnyRef](
     override val returnType: TypingResult
 ) extends CustomLazyParameter[T] {
 
-  def this(
-      compiledParameter: CompiledParameter,
+  override val evaluate: Evaluate[T] =
+    LazyParmeterEvaluator.evaluate(compiledParameter, expressionEvaluator, nodeId, jobData)
+
+}
+
+class SpelTemplateEvaluableLazyParameter[T <: AnyRef](
+    compiledParameter: BaseCompiledParameter,
+    expressionEvaluator: ExpressionEvaluator,
+    nodeId: NodeId,
+    jobData: JobData
+) extends TemplateLazyParameter[T] {
+
+  override val evaluate: Evaluate[T] =
+    LazyParmeterEvaluator.evaluate(compiledParameter, expressionEvaluator, nodeId, jobData)
+
+  override def templateExpression: TemplateExpression = compiledParameter.expression match {
+    case expression: SpelExpression =>
+      expression.subexpressions match {
+        case Some(subexpressions) =>
+          val templateParts = subexpressions.map {
+            case TemplatedExpression(expression) => {
+              new TemplateExpressionPart.TemplatedPart {
+                override val evaluate: Evaluate[String] = context => {
+                  expressionEvaluator.evaluate[String](expression, "expressionId", nodeId.id, context)(jobData).value
+                }
+              }
+            }
+            case NonTemplatedValue(value) => TemplateExpressionPart.NonTemplatedPart(value)
+          }
+          TemplateExpression(templateParts)
+        case None =>
+          throw new IllegalStateException("Non SpEL-template expression received in SpelTemplateLazyParameter")
+      }
+    case _ => throw new IllegalStateException("Non SpEL expression received in SpelTemplateLazyParameter")
+  }
+
+  override def returnType: TypingResult = Typed[String]
+}
+
+private[this] object LazyParmeterEvaluator {
+
+  def evaluate[T <: AnyRef](
+      compiledParameter: BaseCompiledParameter,
       expressionEvaluator: ExpressionEvaluator,
       nodeId: NodeId,
       jobData: JobData
-  ) =
-    this(compiledParameter, expressionEvaluator, nodeId, jobData, compiledParameter.typingInfo.typingResult)
-
-  override val evaluate: Evaluate[T] = { ctx: Context =>
+  ): Evaluate[T] = { ctx: Context =>
     expressionEvaluator
       .evaluateParameter(compiledParameter, ctx)(nodeId, jobData)
       .value
       .asInstanceOf[T]
   }
 
 }
+
+object EvaluableLazyParameterFactory {
+
+  def build[T <: AnyRef](
+      compiledParameter: BaseCompiledParameter,
+      expressionEvaluator: ExpressionEvaluator,
+      nodeId: NodeId,
+      jobData: JobData,
+      parameterDefinition: ParameterDef,
+      typingResult: TypingResult
+  ): LazyParameter[T] = {
+    EditorBasedLanguageDeterminer.determineLanguageOf(parameterDefinition.editor) match {
```

Comment:
Not sure if its the best idea to base on parameters definition editor to determine if `SpelTemplateEvaluableLazyParameter` or standard `EvaluableLazyParameter` should be created. Maybe a better idea would be to add another flag field in `pl.touk.nussknacker.engine.api.definition.Parameter` like `isTemplateParameter` and then throw exception if editor is not a SpelTemplate editor or `isLazyParameter` is set to false.

Reply:
I think that we don't need to have this language base pattern matching at all. We can have a factory method for `EvaluableLazyParameter` where we check what kind of expression is inside `compiledParameter.expression`

Reply:
Damn, somehow didn't see that, thanks :D corrected